_posts/2025-04-19-navigating-pagination-with-hybrid-query.md

```diff
@@ -0,0 +1,225 @@
+
+### Results
+
+Our tests used an index containing 134 million documents (4 copies of the `noaa_semantic_search` dataset).
+
+The following graph presents the results of running a hybrid query containing three BM25 subqueries: term, range, and date.
+
+![Performance-graph-of-hybrid-query-with-pagination](/assets/media/blog-images/2025-04-19-navigating-pagination-with-hybrid-query/Performance-graph-of-hybrid-query-with-pagination.png){:class="img-centered" width="500px"}
+
+The following graph presents the results of running a metric aggregation with a hybrid query containing three BM25 subqueries: term, range, and date.
+
+![metric-aggregation-performance-graph](/assets/media/blog-images/2025-04-19-navigating-pagination-with-hybrid-query/metric-aggregation-performance-graph.png){:class="img-centered" width="500px"}
+
+The following graph presents the CPU utilization results.
+
+![cpu-utilization](/assets/media/blog-images/2025-04-19-navigating-pagination-with-hybrid-query/cpu-utilization.png){:class="img-centered" width="500px"}
+
+### Performance summary
+
+The results show gradual, predictable performance changes:
+- There are no significant latency or CPU utilization spikes.
+- Latency metrics (p50, p90, p99, p100) and CPU utilization increase steadily with increasing pagination depth.
+
+## Best practices for using pagination_depth
+
+Our benchmarking shows no significant performance impact when fetching the second page, even with a `pagination_depth` of 10,000. However, both latency and CPU utilization increase with pagination depth. To optimize performance, consider the following best practices.
+
+### Optimize pagination depth based on data distribution
+
+The efficiency of pagination depends on how evenly your data is distributed across shards. You can check this distribution using the `_cat/shards` API. If the documents are equally distributed among shards, then even a smaller `pagination_depth` can provide more hybridized search results per subquery.
+
+For example:
+* With `pagination_depth = 20` across 3 shards, you can hybridize up to 60 results per subquery.
+* This assumes that each shard contains at least 20 results per subquery (thus, the data is evenly distributed).
+* If there are no duplicate entries in the subqueries, you can paginate through up to 120 hybridized search results.
+
+### Manage uneven data distribution
+
+Custom routing can lead to uneven data distribution across shards, resulting in some shards returning zero results, which reduces the total number of hybridized results.
+
+Additionally, some shards may contain all the relevant results. If a shard contains more than 20 results per subquery, it can still return only 20 results (based on `pagination_depth`). As a result, the coordinator node will receive fewer results for hybridization, reducing the final search result size.
```

Comment:
```suggestion
Additionally, some shards may contain all the relevant results because of custom routing. Continuing the earlier example, if a shard contains more than 20 results per subquery, it can still return only 20 results (based on `pagination_depth`). As a result, the coordinator node will receive fewer results for hybridization, reducing the final search result size.
```